Corgibytes.Freshli.Cli/Functionality/Git/ComputeHistory.cs

```diff
@@ -9,48 +8,109 @@ namespace Corgibytes.Freshli.Cli.Functionality.Git;
-                    CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Sunday)),
-
-            // Grouped by month
-            "month" => commitHistory.GroupBy(commit => commit.CommittedAt.Month),
-
-            // Grouped by year
-            "year" => commitHistory.GroupBy(commit => commit.CommittedAt.Year),
-            _ => null
+            return new List<HistoryIntervalStop>();
+        }
+
+        // Technically this could return null, but not in our case since we already know we have some commits
+        var oldestCommit = gitCommits.MinBy(commit => commit.CommittedAt)!;
+
+        // Here be dragons!
+        // The code that follows looks odd but it's important to remember we are walking back in time.
+        var range = new List<DateTimeOffset>
+        {
+            startAtDate
         };
 
-        // Per group, select the most recent commit.
-        // So if grouped per day, pick the commit committed last on that day.
+        var rangeStartDate = DetermineRangeStartDate(startAtDate, quantifier);
+
+        if (rangeStartDate != startAtDate)
+        {
+            range.Add(rangeStartDate);
+        }
+
+        while (rangeStartDate > oldestCommit.CommittedAt)
+        {
+            var dateTimeOffset = quantifier switch
+            {
+                "d" =>
+                    // Go back at interval -x days
+                    rangeStartDate.AddDays(-interval),
+                "w" =>
+                    // Go back at interval of -7 * interval days
+                    rangeStartDate.AddDays(-7 * interval),
+                "m" =>
+                    // Go back at interval of -x months
+                    rangeStartDate.AddMonths(-interval),
+                "y" =>
+                    // Go back at interval of -x years
+                    rangeStartDate.AddYears(-interval),
+                _ => default
+            };
+
+            if (dateTimeOffset < oldestCommit.CommittedAt)
+            {
+                range.Add(oldestCommit.CommittedAt);
+                break;
+            }
+
+            range.Add(dateTimeOffset);
+            rangeStartDate = dateTimeOffset;
+        }
+
+        // Foreach offset in range, select the youngest commit, as long as it's not younger than the offset.
         return (
-                from groupedHistory in groupedHistories
-                select groupedHistory.MaxBy(i => i.CommittedAt)
-                into mostRecentCommitForThisGroupedPeriod
-                where mostRecentCommitForThisGroupedPeriod != null
-                select new HistoryIntervalStop(mostRecentCommitForThisGroupedPeriod.ShaIdentifier,
-                    mostRecentCommitForThisGroupedPeriod.CommittedAt))
-            .ToList();
+            from offset in range
+            let lastCommitForOffset = gitCommits.First(commit => commit.CommittedAt <= offset)
+            select new HistoryIntervalStop(lastCommitForOffset.ShaIdentifier, offset))
```

Comment:
I'm not a fan of this style of LINQ expressions. Can we please switch to the other style? I wish I knew a way to make a linter rule for this. :)

Reply:
Actually, now that I think about it. I'm okay with this for now. I can create an issue for us to add a linter rule for this instead.